### ISSUE TITLE
Feature request: ban createRef in function components

### ISSUE THREAD

**Issue by danvk** [yannickcr/eslint-plugin-react#2966]
Using `createRef` in a function component is almost always a mistake because it will create a new ref on every render. Usually you want `useRef` (see [this discussion][1]). Since this is such an easy mistake to make, it would be helpful to have a linter rule to flag it.

[1]: https://stackoverflow.com/questions/54620698/whats-the-difference-between-useref-and-createref/54620836#54620836

**Comment by ljharb** [comment]
That seems like a good rule for eslint-plugin-react-hooks. This plugin can’t assume you have hooks available, so such a rule wouldn’t be useful for react versions without them.

**Comment by danvk** [comment]
Thanks @ljharb. I filed an issue against the React repo, we'll see what they say: https://github.com/facebook/react/issues/21256

**Comment by bvaughn** [comment]
IMO this does seem like something that should be warned about, but it also seems odd for `eslint-plugin-react-hooks` to warn about it (since those rules are all about hooks and `createRef` is not a hook).

Regardless of whether the version of React in question supports hooks, using `createRef` in a function body like this is always a mistake, right?

**Comment by ljharb** [comment]
Not necessarily - it could be memoized manually, or used inside another memoizing hook.

**Comment by bvaughn** [comment]
Not when it's in the top level body of the function component (not within a conditional) as shown in the linked example.

**Comment by ljharb** [comment]
(warning: off the top of my head) Something like: 
```jsx
function Foo({ a }) {
  const ref = useMemo(() => createRef(), [a]);
  const fn = useCallback(() => doSomethingWith(ref?.current), [ref]);
  return <Bar onSomething={fn} />;
}
```

Obviously you might prefer `useRef` here instead of `useMemo` - but that doesn't mean there's something *wrong* with this usage.

**Comment by bvaughn** [comment]
> Obviously you might prefer `useRef` here instead of `useMemo` - but that doesn't mean there's something _wrong_ with this usage.

Kind of does 😄  Seems to suggest (to me) some fundamental misunderstanding of how the APIs work.

**Comment by danvk** [comment]
@ljharb that usage of `createRef` seems fine and @bvaughn's caveat about "in the top level body of the function component" would prevent a false positive on it.

**Comment by ljharb** [comment]
That’s all a reasonable position to take.

Unfortunately, eslint-plugin-react-hooks doesn’t export (or depend on a package for) the hook detection logic, which means we have to copy-paste it, which seems very unfortunate.

@bvaughn even though i really do think any rule dealing with hooks whatsoever belongs in the “react hooks” eslint plugin (that’s the implied position you occupied by publishing a so-named eslint plugin, imo), if you think it would be better in this plugin, then could at least the hook logic be factored out so we can reuse it?

**Comment by bvaughn** [comment]
cc @gaearon who may have thoughts here. (He's the primary owner of the hooks lint rule.)

**Comment by eirikur-grid** [comment]
I've resorted to adding yet another plug-in, https://github.com/rel1cx/eslint-react, just so that I can enable its [no-create-ref](https://eslint-react.xyz/rules/no-create-ref) rule.
